Add description of square bracket parsing algorithgm to README. Plus update the readme with description of how crossmatch is calculated in those cases. (#1171)
mild-blue/txmatching: documentation/README.md

* "beta chain" refers to the second code of a double antibody.
* "positive MFI" is the MFI value that is equal to or above the cutoff.
* "negative MFI" is the MFI value below the cutoff.

We parse each antibody one by one and detect several situations:

#### 1. Antibody chain is negative

```text
Example:
DP*[01:01;02:02] MFI 100, cutoff 2000
DP*[01:01;03:03] MFI 200, cutoff 2000
DP*[03:03;02:02] MFI 3000, cutoff 2000
```

In this case, the chain `DPA1*01:01` is negative because there is no
representation
of this chain with MFI above the cutoff among all antibodies.
As a result, we parse this chain as:

```text
DPA1*01:01 MFI 150 (average MFI), cutoff 2000
```

Please note that the chain `DPB1*02:02` is not negative because it has a
positive
representation in the antibody `DP*[03:03;02:02] MFI 3000, cutoff 20000`.
Therefore, this chain is mixed.

Similarly, the chain `DPB1*03:03` is negative and parsed in the same way
as `DPA1*01:01`.

#### 2. Both chains are only positive

```text
Example:
DP*[03:03;02:02] MFI 3000, cutoff 2000
DP*[01:01;03:03] MFI 200, cutoff 2000
DP*[03:03;01:01] MFI 2100, cutoff 2000
DP*[04:01;03:03] MFI 3000, cutoff 2000
```

> In this case, the antibody `DP*[03:03;02:02]` has both chains positive because
> either `DPA1*03:03` or `DPB1*02:02` have only positive MFIs among all

Reviewer: ```suggestion
In this case, the antibody `DP*[03:03;02:02]` has both chains positive because
both `DPA1*03:03` and `DPB1*02:02` have only positive MFIs among all
```

antibodies.
As a result, we parse this antibody as two separate chains:

```text
DPA1*03:03 MFI 2550 (average MFI), cutoff 2000
DPB1*02:02 MFI 3000 (average MFI), cutoff 2000
```

The antibody `DP*[03:03;01:01]` is also only positive.
However, `DP*[04:01;03:03]`
is not positive because it has negative representation of the
chain `DPB1*03:03` in the antibody `DP*[01:01;03:03]`.

#### 3. One chain is only positive, another one is mixed:

```text
Example:
DP*[03:03;02:02] MFI 3000, cutoff 2000
DP*[01:01;02:02] MFI 100, cutoff 2000
DP*[04:01;01:05] MFI 3000, cutoff 2000
DP*[03:01;01:05] MFI 2500, cutoff 2000
DP*[01:01;01:05] MFI 200, cutoff 2000
DP*[03:01;06:02] MFI 100, cutoff 2000
```

> In this case, the antibody `DP*[03:03;02:02]` has only a positive chain,
> `DPA1*03:03`, and a mixed chain, `DPB1*02:02`, because `DPB1*02:02` has a

Reviewer: ```suggestion
In this case, the antibody `DP*[03:03;02:02]` has an only positive chain,
`DPA1*03:03`, and a mixed chain, `DPB1*02:02`, because `DPB1*02:02` has a
```

negative
MFI representation in the antibody `DP*[01:01;02:02]` with an
MFI of 100 (below the cutoff of 2000), while `DPA1*03:03` does not.
As a result, we parse the positive chain `DPA1*03:03` with its average MFI among
all antibodies with this chain:

```text
DPA1*03:03 MFI 3000 (average MFI), cutoff 2000
```

This mixed chain `DPB1*02:02`, which has positive
representation just in this antibody `DP*[01:03;02:02]` (there aren't any other
antibodies where this chain has positive MFI), is parsed as

```text

> DPB1*02:02 MFI 100 (average negative MFI), cutoff 2000

Reviewer: Napadá mě tady ještě zmínit že to děláme protože předpokládáme že positive MFI je způsobeno tím druhým chainem, takže tuto positivni hodnotu pro ten mixed chain neuvažujeme.

```

Pay attention that a similar mixed chain, `DPB1*01:05`,
but for the antibody `DP*[04:01;01:05]`, is not parsed through this antibody but
through the `DP*[03:01;01:05]` antibody, as explained in the next section.
In summary, if we encounter this positive and mixed chain case where the mixed 
chain has several positive representations among all the antibodies, 
we expect to parse the mixed chain through other antibodies.

#### 4. Both chains are mixed

```text
Example:
DP*[03:03;02:02] MFI 3000, cutoff 2000
DP*[01:01;02:02] MFI 100, cutoff 2000
DP*[03:03;04:01] MFI 200, cutoff 2000
DP*[03:03;05:02] MFI 2200, cutoff 2000
```

In this case, the antibody `DP*[03:03;02:02]` has both chains as mixed because
its chains,
`DPA1*03:03` and `DPB1*02:02`, have negative representation in the antibodies

> `DP*[03:03;04:01]` and `DP*[01:01;02:02]`, respectively.

Reviewer: ```suggestion
`DP*[03:03;04:01]` and `DP*[01:01;02:02]` respectively.
```

```text
Example:
DP*[03:03;02:02] MFI 3000, cutoff 2000
DP*[01:01;02:02] MFI 100, cutoff 2000
DP*[04:01;01:05] MFI 3000, cutoff 2000
DP*[03:01;01:05] MFI 2500, cutoff 2000
DP*[01:01;01:05] MFI 200, cutoff 2000
DP*[03:01;06:02] MFI 100, cutoff 2000
```
In this case, the antibody `DP*[01:03;02:02]` has only a positive chain, 
`DPA1*03:03`, and a mixed chain, `DPB1*02:02`, because `DPB1*02:02` has a negative 
MFI representation in the antibody `DP*[01:01;02:02]` with an 
MFI of 100 (below the cutoff of 2000), while `DPA1*03:03` does not.
As a result, we parse the positive chain `DPA1*03:03` with its average MFI among 
all antibodies with this chain:
```text
DPA1*03:03 MFI 3000 (average MFI), cutoff 2000
```

This mixed chain `DPB1*02:02`, which has positive
representation just in this antibody `DP*[01:03;02:02]` (there aren't any other
antibodies where this chain has positive MFI), is parsed as 
```text
DPB1*02:02 MFI 100 (average negative MFI), cutoff 2000
```

Pay attention that a similar mixed chain, `DPB1*01:05`, 
but for the antibody `DP*[04:01;01:05]`, is not parsed through this antibody but 
through the `DP*[03:01;01:05]` antibody, as explained in the next section. 

> In summary, if there are several positive representations in this positive + mixed case, 

Author: Tady to nese smysl, o kterém diskutujeme tady https://github.com/mild-blue/txmatching/pull/1171/files/16ff94cea7a8f074b86cb01af72ee6e1723a7cc6#r1183367165, tak korektní asi bude varianta vyplývající z našich dvou:
```suggestion
In summary, if we detect the mixed chain, which has several positive representations among all the antibodies, 
in the positive + mixed double antibody, 
```
Souhlas? Nebo vadí takový popis "positive + mixed" v non tech README?

> In summary, if there are several positive representations in this positive + mixed case, 

Reviewer: Ten odkaz nefunguje :( . Nicméně koukám na tvůj návrh a stále mi to přijde matoucí kvůli pořadí těch vět. Mě to dává smysl třeba takhle:

```
In summary, if we encounter this positive and mixed chain case where the mixed chain has several positive 
representations among all the antibodies, we expect to parse the mixed chain through other antibodies.
```

> we expect to parse this mixed chain through other antibodies.

Author: V tomhle případě ne, ale nejčastěji čekáme, že tu mixed s několika positive parsneme v mixed+mixed case.
Viz test, který jsme přidal, protože jsem nějak nenašel existující test pro tento case:
https://github.com/mild-blue/txmatching/blob/863842ec395f4be55f9b958102a74170c6255a19/tests/utils/hla_system/test_code_parser.py#L310-L351

If at least one of the chains is not in the `parsed` list yet, we examine the antibody with the following logic:
</br>
First, we check if the antibody has a positive MFI.

#### Double antibody has a positive MFI:
If **both antibody chains have only positive MFI representation** among all the antibodies 
in the list for parsing, we add each chain **separately** to the `parsed` list with the arithmetic mean of the MFI from all antibodies with this chain in the list.</br></br>
If **one of the chains has at least one negative MFI representation** and the **other chain has only positive MFI representation** among other antibodies in the list, we add the **chain with only positive MFI representation** to the `parsed` list with the arithmetic mean of the MFI from all antibodies with this chain. For the **chain with mixed MFI representation**, if this is the only positive occurrence of this chain, we add it to the `parsed` list with the arithmetic mean of the MFI from all antibodies with negative MFI for this chain in the list. However, if there is another positive occurrence for this chain, we skip it.</br></br>
If **both chains have at least one negative MFI among other antibodies**, 

> we add them to the `parsed` list as one double antibody with the provided MFI. Moreover, we add each chain **separately** to the `parsed` list, and we denote their type as "theoretical", with the arithmetic mean of the MFI from all antibodies with this chain in the list.

Reviewer: Tady se bere jenom mean pres positive mfi ne?

Author: Jo, jen jsem ještě to neopravil. Pokud náhodou děláš review (zvlášť u toho nového readme netechnického), tak asi počkej, protože zatím je to draft. Hodně věcí se tam mohou změnit, abys to neprocházel dvakrát

Reviewer: Procital jsem si to parsovani kvuli jinemu review, tak jsem si tohohle všiml a nechtěl jsem na to zapomenout :) . S review tohoto počkám na request.

## Square bracket antibody parsing algorithm

When we receive antibodies in notation DP*[01:01;02:02] we are using a special algorithm to deduce whether there are
antibodies against both alpha and beta alleles or just one of them.

### Algorithm description
Antibodies enclosed in square brackets are called double antibodies.

Let us firstly explain the terminology we used to describe the algorithm:

* The alpha chain refers to the first code of a double antibody.
* The beta chain refers to the second code of a double antibody.
* Positive MFI is the MFI value is equal to or above the cutoff.
* Negative MFI is the MFI value below the cutoff.

At the beginning of the algorithm execution, we have three lists:
1. antibodies - a list of all antibodies to parse.
2. parsed - a list of parsed antibodies.

> 3. parsing_issues - a list of parsing issues that occur during the algorithm running.

Reviewer 2: jaj toto som myslela ze "algorithm running" -> "algorithm execution"

We parse each antibody in the list one by one using a for-loop. If an antibody is 

> already in the parsed list, we move on to the next antibody. 

Reviewer 2: tu by som este highlightla ze if both chains are already parsed

> First, we check if the antibody has a positive MFI.

Reviewer 2: tu by som tiez highlightla ze teda ak neni sparsovane, tak ideme sem

#### Double antibody has a positive MFI:
If **both antibody chains have only positive MFI representation** among all the antibodies 
in the list for parsing, we add each chain **separately** to the parsed list, 
with the arithmetic mean of the MFI from all antibodies with this chain in the list.</br>
If **one of the chains has at least one negative MFI representation** and the **another one has 
only positive MFI representation** among other antibodies 
in the list, we add its chain to the parsed list with the arithmetic mean of the MFI 
from all antibodies with negative MFI for this chain in the list, and add the other chain with the 

> arithmetic mean of the MFI from all antibodies with this chain, as in the first case.</br>

Reviewer 2: If one of the chains has at least one negative MFI representation and the other chain has only positive MFI representation among other antibodies in the list, we add the chain with only positive MFI representation to the parsed list with the arithmetic mean of the MFI from all antibodies with this chain. For the chain with mixed MFI representation, if this is the only positive occurrence of this chain, we add it to the parsed list with the arithmetic mean of the MFI from all antibodies with negative MFI for this chain in the list. Otherwise, if there is another positive occurrence for this chain, we skip it for now.

taketo nieco, asi este treba highlightnut dake slova a ak sa to da napisat krajsie tak prepisat

If **both chains have at least one negative MFI among other antibodies**, 
we add them to the parsed list as one theoretical double antibody with the arithmetic 
mean of the MFI from all antibodies with these chains in the list. 
Also we add a parsing issue about this to the parsing issue list.
Then we add each chain **separately** to the parsed list if they are not already present, 

> with the arithmetic mean of the MFI from all antibodies with this chain in the list.

Reviewer 2: este stale nie tak. v tomto pripade sa deje toto:

if both chains have positive and negative mfis:
1. add double antibody of type NORMAL with THIS MFI
2. raise parsing issue
3. add both chains separately of type THEORETICAL with AVERAGED MFI

#### Double antibody has a negative MFI:
If a chain has **only negative MFI representation** among other antibodies in the list, 
we add this chain to the parsed list with the arithmetic mean of the MFI from all 
antibodies in the list. Otherwise, we skip this antibody and move on to the next one.

> 

Reviewer 2: tuto sa to riesi separatne pre chainy, v zmysle - ak je chaina len negative, tak ju pridam, ak je chaina aj positive tak ju nepridam. ale z tohto textu mam pocit akoby ked jedna chaina ma daky positive vyskyt tak skipujem cele antibody

Reviewer 2: proste len ta posledna veta by ma asi zmiatla

When we receive antibodies in format DP*[01:01;02:02] we are using a special algorithm to deduce whether there are
antibodies against both alpha and beta alleles or just from one of them.

### Algorithm description
Antibodies enclosed in square brackets are called double antibodies.

Let us first explain the terminology we used to describe the algorithm:

* The alpha chain refers to the first code of a double antibody.
* The beta chain refers to the second code of a double antibody.
* Positive MFI is the MFI value above the cutoff.
* Negative MFI is the MFI value below the cutoff.

At the beginning of the algorithm, we have three lists:
1. antibodies - a list of all antibodies to parse.
2. parsed - a list of parsed antibodies.

> 3. parsing_issues - a list of parsing issues that occur during the algorithm running.

Reviewer 2: algorithm execution
